src/model.jl

```diff
@@ -1,3 +1,20 @@
+""" Abstract type for an abstract model formulated in the state space form
+"""
+abstract type AbstractStateSpaceModel <: AbstractMCMC.AbstractModel end
+
+struct Trace{F,R<:TracedRNG}
+    f::F
+    ctask::Libtask.CTask
+    rng::R
+end
+
+struct SSMTrace{F<:AdvancedPS.AbstractStateSpaceModel,R<:TracedRNG}
+    f::F
+    rng::R
+end
+
+const Particle = Union{Trace,SSMTrace}
```

Comment:
This does not seem very extendable. And leads to `Particle` being an abstract type instead of an alias of `Trace`. This seems dangerous both for internal implementations and will likely break downstream code that used the alias.

I think a better alternative might be to change `Trace` to
```julia
struct Trace{F,R<:TracedRNG}
    f::F
    rng::R
end
```
and integrate `ctask` in `f` (probably breaking). Then one could just dispatch on type parameter `F`, and e.g. define
```julia
const SSMTrace{R} = SSMTrace{<:AbstractStateSpaceModel,R}
```
(an abstract type!) as alias for dispatching.

Reply:
Agree, missed the issue with downstream using `Particle`. I think unifying `Trace` as you said is probably the best option. It would change a few things in the rest of the package though, not just pgas.

Reply:
@FredericWantiez I agree with @devmotion's suggestion here. It is better to unify these `Trace` types.

Reply:
I moved verything in `model.jl` and unified `Trace`:
https://github.com/TuringLang/AdvancedPS.jl/blob/1f152785a42739afd98d84ebd889b3bde88ef24a/src/model.jl#L1-L34